Packet lights

jordan-powers opened the pull request:
This works, and is much better, so we can merge it back into the main Blinky_Lights branch

terbos90803 committed:
First implementation of packet-based painting protocol.

jordan-powers committed:
Java using Packets

jordan-powers committed:
Updated animations

terbos90803 committed:
Refactor and cleanup pass.  Combine all paint commands to use paintPattern.

jordan-powers committed:
Brigtness only applied to packet. Increased clarity

jordan-powers committed:
Merge branch 'Packet_Lights' of https://github.com/momentumfrc/2017Steamworks into Packet_Lights

jordan-powers committed:
Sync only every so often

jordan-powers committed:
Bug fixes

jordan-powers committed:
Comments, part 1

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
-		this.brightness = brightness;
 		
+		// Needs to be daemon so that it doesn't block the exit of the JVM
 		this.setDaemon(true);
```
What's your reasoning for setting this?  When does the JVM on the robot ever exit?
If the JVM is killed, all threads will be killed too.

The key difference between a daemon thread and non-daemon is that if you have two normal threads running, the JVM will stay running until both exit.  If one is a daemon, only the normal one is required to exit before the JVM ends.  The daemon thread would just get killed at that point.

I'm not convinced that setting this thread as a daemon is necessary or appropriate.  I'm open to contrary arguments.

jordan-powers replied:
Yeah, I guess if we just power off the robot, it doesn't matter what threads are daemon or not, they're all getting killed.
I'll remove this line

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
-		for(int i = 0; i < out.length; i++) {
-			out[i] = dim(out[i]);
+	public void setAnimation(Animation newAnimation) {
+		synchronized (current) {
```
In thinking about this some more, I don't think we can sync on current since it's getting replaced within the block.  I'm also not convinced that we need to sync on changing current since we can replace it cleanly in the class as long as the executor method (run()) makes a local copy for the critical section.

jordan-powers replied:
Okay, that makes sense. I'll remove this synchronization an have the run() method copy the animation instead

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
+		}
+		// Tells the thread to continue. 
+		synchronized (pauseLock) {
+			pauseLock.notifyAll();
```
Good thought to use a wait here, although I wouldn't strictly require it.  The animation thread should have low enough impact on the system that if it woke up periodically to discover that there's no current animation (which never happens), that wouldn't be a problem.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
+		while(!Thread.interrupted()){
+			
+			int delay;
+			synchronized (current) {
```
Again, probably a better solution instead of sync on current would be to make a local copy of the reference to current and use it in this scope.  Then if the member reference gets replaced, it'll simply get picked up on the next loop.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
+				// show current frame. keep track of how long it took
+				timeToSend = out.show(current.animate());
+				if(timeToSend < 0) {
+					System.out.println("Failed to write to neopixels, suspending animation thread");
```
What's your justification for suspending the animation if there's a failure?  Why is this only detected by the negative return value?
Two suggestions here:
1. out.show() should not be responsible for calculating the time to execute since it has no idea how long current.animate() took.  If you care about the total time, it should all be calculated locally.
1. failures to generate or send the animation should be handled as exceptions.  Putting this in a try/catch block makes much more sense for error handling since the vast majority of the time it works correctly.

jordan-powers replied:
Yeah, I should definitely be calculating the time to send in the Animator thread. I was returning the value instead because I had been calculating the time to send as part of debugging the I2C code, and I hadn't thought I could just calculate the time in the Animator thread. I'll remove the return value, and have the try/catch block just log the error to Sys.err if there is an error.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
+				if(timeToSend < 0) {
+					System.out.println("Failed to write to neopixels, suspending animation thread");
+					try {
+						pauseLock.wait();
```
I'm uncomfortable with a potentially infinite wait here.  It should at least have a timeout or should simply be a sleep.

jordan-powers replied:
I'll remove this whole if statement, the errors will be caught in the NeoPixels class.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
+				} else {
+					//System.out.format("Send took %dms, delay requested is %dms\n", timeToSend, delay);
+					delay -= timeToSend;
+					delay = (delay < 0) ? 0 : delay;
```
This line is useless.  The opposite condition is tested on the next line.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,110 @@
-			animate = new AnimatorThread(pixels, currentState, currentAnimation, masterBrightness);
-			animate.start();
+		if(newAnimation == null) {
+			System.out.println("Can't set a null animation!!");
```
Perhaps setting null should choose the Solid Black animation.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Color.java:
```diff
@@ -1,30 +1,97 @@
 public class Color {
 	private int r,g,b;
 	
+	private static double brightness = MoPrefs.getInstance().getBrightness();
```
You're picking this up directly from MoPrefs here, yet the only way to change it is to use the static setter below.  This seems awkward.

Does MoPrefs support the Observer pattern where updates can be pushed to interested parties?  If not, then you might consider a few options:
1. Have an updateBrightness() method that pulls the latest value from MoPrefs.
1. Decouple this class from MoPrefs entirely and *always* pass in the brightness that you really want.
1. Decouple brightness from the Color class entirely.  Color exists as an immutable thing.  If you need to filter a color in any way, run it through a filter class, e.g. BrightnessFilter that takes a Color and returns a Color.

jordan-powers replied:
I'll go for option 3. I'll have the BrightnessFilter implement the listener interface and I'll register it to listen to changes to the brightness value.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Color.java:
```diff
@@ -1,30 +1,97 @@
+ *
+ */
 public class Color {
 	private int r,g,b;
```
Since Color is immutable, these should be final.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Display.java:
```diff
@@ -1,7 +1,17 @@
+	 * @param pixels
+	 * @return Time taken to show the data
+	 */
+	long show(Packet[] pixels);
```
I would probably choose a different name than pixels here.  An individual paint command can represent multiple pixels and/or the show command.  Perhaps a better name would be 'commands'.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -25,103 +29,30 @@ public static NeoPixels getInstance() {
 	
 	private NeoPixels() {
-		millis = System.currentTimeMillis();
 		strip = new I2C(I2C.Port.kOnboard, 16);
```
Please document the magic number 16.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -25,103 +29,30 @@ public static NeoPixels getInstance() {
+			long millis = System.currentTimeMillis();
+			// Send a sync packet every SYNC_FREQ frames
+			if(syncidx == SYNC_FREQ) 
+				strip.writeBulk(Packet.syncPacket(), 16); // Send synchronize packet
```
Magic number 16.  This requires the code to understand what number is needed for the packet type.  Much better to ask the Packet, e.g. Packet.getSyncPacketSize().

terbos90803 replied:
An even better idea here would be to make strip into a NeoPixelsIO class with a method for sending.

```java
class NeoPixelsIO {
 void send(Packet packet) {}
}
```

or something like that.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -1,22 +1,26 @@
 	private static NeoPixels instance;
 	
+	private final int SYNC_FREQ = 1000;
+	private int syncidx = 1000;
```
init to 0

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -25,103 +29,30 @@ public static NeoPixels getInstance() {
+			// Send a sync packet every SYNC_FREQ frames
+			if(syncidx == SYNC_FREQ) 
+				strip.writeBulk(Packet.syncPacket(), 16); // Send synchronize packet
+			syncidx = (++syncidx > SYNC_FREQ) ? 0 : syncidx;
```
not necessary.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -25,103 +29,30 @@ public static NeoPixels getInstance() {
-				strip.write(1, safe(color.getBlue()));
+			long millis = System.currentTimeMillis();
+			// Send a sync packet every SYNC_FREQ frames
+			if(syncidx == SYNC_FREQ) 
```
Better written as:
if (++syncidx >= SYNC_FREQ) {
  strip.writeBulk()
  syncidx = 0;
}

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -25,103 +29,30 @@ public static NeoPixels getInstance() {
 		} catch (Exception e) {
-			System.out.println(e.getMessage());
-			System.out.println(e.getStackTrace());
+			// The generic try-catch prevents an error in the purely cosmetic neopixels from killing the whole robot
```
It should be better to throw a Checked exception instead of returning -1 since I'm not convinced that this method should be returning anything.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Packet.java:
```diff
@@ -0,0 +1,162 @@
+public class Packet {
+	// The max size of a packet is 16 bytes
+	// Before a packet is sent, B_BUFFER is filled with that packet's data. B_BUFFER is then passed to the I2C writeBulk() method
+	private static final ByteBuffer B_BUFFER = ByteBuffer.allocateDirect(16);
```
The docs for ByteBuffer warn against allocating direct buffers unless you really need them.  Why did you choose direct?

In fact, why did you choose ByteBuffer over just a plain java array?  It seems you had to do a lot of extra work dealing with the ByteBuffer that wouldn't be necessary with the array.

jordan-powers replied:
The ByteBuffer needs to be direct because if you look at the implementation of the I2C class, it requires a direct ByteBuffer, and throws an IllegalArgumentException() if the ByteBuffer is not direct.
I'm using ByteBuffer because if you look at the implementation of the I2C class, when you pass writeBulk() an array of bytes it creates a new direct ByteBuffer to hold them. I figured that it would be more memory efficient to just have one direct ByteBuffer that is read from rather than to be constantly creating new ByteBuffers, since direct buffers 'typically have somewhat higher allocation and deallocation costs than non-direct buffers' and it's recommended that 'direct buffers be allocated primarily for large, long-lived buffers'

terbos90803 replied:
That makes a lot of sense when you explain it. I'm a little (a lot) disappointed in WPI for providing an API that's poorly documented in a way that could cause significant problems.

That said, since you have a good reason to use ByteBuffer, you should still be following some data encapsulation rules.  I would not have the singleton ByteBuffer be part of the Packet class, but part of the NeoPixel device class, or another class that wraps/extends the I2C base class.  That way, the details of Packet live only in Packet and the details of the I2C implementation live only there.

jordan-powers replied:
Okay, I subclassed the I2C into NeoPixelIO, like you suggested earlier, so I'll make the buffer a field on that object

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Packet.java:
```diff
@@ -0,0 +1,162 @@
+	// All according to specification
+	private final byte DISPLAY_FRAME = 0x01;
+	private final byte SET_SINGLE = 0x02;
+	private final byte SET_AREA = 0x03;
```
AREA infers a 2-dimensional thing.  A better choice would be SERIES or RUN.  I personally prefer RUN since we used that all the time in our imaging code at Xerox.  You'll also find that term used in Run Length encoding, an early compression technique where long sequences of identical bytes are replaced with a value/count pair.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Packet.java:
```diff
@@ -0,0 +1,162 @@
+	/**
+	 * Creats a DISPLAY_FRAME packet
+	 */
+	private Packet() {
```
This could be handled a lot more cleanly.  My recommendation would be to use the Factory pattern to generate Packets.  The Packet class itself shouldn't have any knowledge of how to make specific types of Packets, nor should particular constructor signatures magically map to commands.

I can help flesh this out if you're not familiar with this pattern.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/animations/Animation.java:
```diff
@@ -1,14 +1,14 @@
 	 * @return The new state of pixels
 	 */
-	Color[] animate(Color[] pixels);
+	Packet[] animate();
```
Perhaps a more clear name could be getNextFrame(), but that's not critical.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/animations/Animation.java:
```diff
@@ -1,14 +1,14 @@
-	Color[] animate(Color[] pixels);
+	Packet[] animate();
 	/**
 	 * Gets time to wait before calling animate. Returns -1 to indicate animate() should only be called once
```
Could also be called getFrameMilliseconds().

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/animations/AnimationSequence.java:
```diff
@@ -28,13 +28,13 @@ public AnimationSequence(Animation[] animations, int[] times) {
+	public Packet[] animate() {
 		if(timepassed >= times[currentidx]) {
 			timepassed = 0;
 			currentidx++;
```
Better written as:
currentidx = (currentidx + 1) % times.length;

I'm also struggling to understand the logic behind timepassed.  I think I see where you're heading, but there's quite a lot of awkwardness around how you're getting there.  This could also be a root cause of some of the glitches we were seeing on the looping.

jordan-powers replied:
I was having the class add up the individual delays to find out how long the animation had been running for, then switching animation if the time running was >= the requested run time of the animation. It makes more sense, though, to store the time when the animation was started, and subtract it from the current time to see how long an animation has been running for.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/animations/Blink.java:
```diff
@@ -27,12 +28,12 @@ public Blink(Color[] colors, int[] waittimes) {
-			out[i] = colors[idx];
+	public Packet[] animate() {
+		Packet out[] = {new Packet(0, colors[idx], 1, 1)};
+		idx++;
```
As always, this pattern is better written as:
idx = (idx + 1) % colors.length;

Here and in all other places.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/animations/Snake.java:
```diff
@@ -55,15 +56,16 @@ public Snake(Color[] snakes, int msBetweenFrames) {
-		offset = (offset < 0) ? 0 : snakes.length - 1;
+		
+		offset = (offset + increment + snakes.length) % snakes.length;
+		// System.out.format("Offset: %d\n", offset);
```
Remove commented code.

terbos90803 reviewed (changes requested):
Welcome to your first professional code review.  There are a lot of comments here, some of which I'd like good answers on before we proceed.  Keep in mind that code reviews are an opportunity for both parties to learn something along the way.  I have already while researching some of my questions.

jordan-powers committed:
WIP Implementing Terry's changes

I need to switch computers, so I'm commiting and pushing. This is definitely still a WIP, though

jordan-powers committed:
WIP 2 Still implementing suggestions

I need to get started on homework, and I need to switch computers again, so I'm pushing and I'll continue working after homework is finished

jordan-powers committed:
Finished Terry's changes

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,97 @@
+			synchronized (pauseLock) {
+				if(delay < 0) {
+					try {
+						pauseLock.wait(WAIT_TIMEOUT);
```
We should probably talk through how you're intending to use the pauseLock.  Is it for animations that are not supposed to be changing?  If so, you're probably better off refreshing that frame occasionally just in case there was a transmission error.  There's very little cost to sending the frame, particularly if it's a solid color.  Thus I'm not seeing the value in supporting the lock when it could just as easily be a sleep(500) to get the same result.

jordan-powers replied:
Yeah, that makes sense. I hadn't considered the benefit of a refresh in case of a transmission error. I'll put a constant delay on the solid class and have it refresh every half-second or so

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,97 @@
+		while(!Thread.interrupted()){
+			// Note how long the send takes
+			long millis = System.currentTimeMillis();
+			Animation animation = current;
```
Probably want to comment why you're making this copy since it may not be obvious to the next reader.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java:
```diff
@@ -6,106 +6,97 @@
+				}
+			}
+			
+			delay -= (System.currentTimeMillis() - millis);
```
On the ride home today, I was thinking through how I might do the frame timing.  The biggest difference I came up with was that the technique you're using is always relative to the last frame.  An alternative would be to reference absolute time (or at least absolute time MOD the total animation loop time) in order to keep the frames on a fixed time rhythm regardless of glitches or interruptions.  But I'm not sure if the rewrite is worth the trouble since this method seems good enough.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Packet.java:
```diff
@@ -0,0 +1,153 @@
+	 */
+	public Packet(int address, Color color) {
+		command = SET_SINGLE;
+		Color dimColor = BrightnessFilter.dimColor(color);
```
Better, but I still think that the BrightnessFilter doesn't belong in the Packet class.  Packet has to do with packaging the command and data for transport, it shouldn't know anything about the data it's packaging, nor how other layers will pass the package along.  This is part of why I was suggesting the use of the Factory pattern to generate packets which separates the Packet from the packet generation.  I'm happy to demonstrate.

jordan-powers replied:
I'm not quite sure what you mean by 'the Factory pattern'? I do agree that the Packet class probably should just be holding data, and not modifying it. Do you mean another class which has different methods that generate Packet objects?

terbos90803 replied:
Here's the definition: https://en.wikipedia.org/wiki/Factory_method_pattern

Although it goes a bit abstract in the middle, so you're probably looking for a more concrete example.

In our specific case, there are two common ways to get this done.  You can have one class that knows how to properly generate other classes, including selection of what specific subclasses should be used for what situations.  If you go that route, I'd be looking for a series of APIs like this:
* Create a command packet:  
`Packet command = Commands.paintPixel(0, color);`  
`Packet command = Commands.paintPixels(0, color, 1);`  
`Packet command = Commands.paintPixelsWithStride(0, color, 1, 1);`

An alternative would be to simply use Packet as the base class for all Packet types.  Then the different commands would all be subclassed off of that like this:
* `Packet command = new PaintPixel(0, color);`
* `Packet command = new PaintPixels(0, color, 1);`
* `Packet command = new PaintPixelsWithStride(0, color, 1, 1);`

It may seem simpler at this time to choose the direct construction method, and that's likely to be the case.  The thing to watch out for is if there is any specific knowledge needed to create the painting commands, then a factory becomes much more useful since it's the central repository for knowledge of how to make the things.

terbos90803 reviewed (commented):
Looking much better.  Just a couple more tweaks and we can move on to the next project.

jordan-powers committed:
Move ByteBuffer to NeoPixelIO

jordan-powers committed:
Removed support for negative delays. Added more comments

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -8,19 +8,24 @@
 	
 	public boolean writePacket(Packet packet) {
-		return writeBulk(packet.fillBuffer(), packet.getPacketSize());
+		packet.fillBuffer(buffer);
```
Knowledge of the existence and necessity for the ByteBuffer lies solely within the NeoPixelsIO class.  There is no reason that the Packet class should ever know anything about the ByteBuffer.  Move the implementation of fillBuffer here.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/NeoPixels.java:
```diff
@@ -8,19 +8,24 @@
-		ByteBuffer packet = Packet.syncPacket();
-		return writeBulk(packet, packet.capacity());
+		Packet.syncPacket(buffer);
+		return writeBulk(buffer, buffer.capacity());
```
Why is the sending of the SyncPacket any different than sending a ShowPacket, or any other packet?  The IO class should only know about generic Packet and how to bundle it into a ByteBuffer and write it to I2C.

I understand the desire to have pre-made Sync and Show packets lying around, but they should be owned outside the IO class.  Perhaps in the NeoPixels class.  But they should be sent via the writePacket method just like any other packet.

jordan-powers replied:
If we use a Factory to make Packets, then the only function of the Packet is to hold a series of bytes, and to calculate the size of that series (so I might just use an array and delete Packet entirely). Then the writePacket method just puts the size as the first byte in the ByteBuffer and fills the rest of the ByteBuffer with the data held by the Packet. Since the maximum size of a packet (including the wrapper) is 16 bytes, that's what I'm using as the size of my ByteBuffer.
However, the structure of a Sync packet is different from that of all other packets. It doesn't have the size byte, it's just 16 (or more) bytes of 0xFF. So I want to fill my entire ByteBuffer with 0xFF. The three approaches I can think of are:
1. Have writePacket method detect when the packet is a Sync packet and not write the size byte
1. Put a separate method on NeoPixelsIO that to send a Sync packet that just fills the ByteBuffer with 0xFF
1. Make the ByteBuffer 17 bytes long and have the packet data be 16 bytes of 0xFF.

I elected to go with the second option, but if you have a better idea I'm all ears

jordan-powers replied:
Now that I think about it, I could have the Factory just return byte arrays for the NeoPixelsIO to load into the buffer, and remove Packet entirely. Then, the factory would be in charge of setting the size byte, and it could have a method to create a synchronization packet that would just be a byte array of length 16 filled with 0xFF.

terbos90803 commented on 2017 Robot/src/org/usfirst/frc/team4999/lights/Packet.java:
```diff
@@ -123,29 +120,27 @@ public byte getCommand() {
-		if(data.length > 0) B_BUFFER.put(data);
-		B_BUFFER.rewind();
-		return B_BUFFER;
+	public void fillBuffer(ByteBuffer buffer) {
```
As per other comment, should move to IO class.

jordan-powers committed:
Use Commands Factory

jordan-powers committed:
Better names and better documentation

terbos90803 reviewed (approved):
Ship It!
